README.md
# rest-firebase [![Build Status](https://secure.travis-ci.org/CodementorIO/rest-firebase.png?branch=master)](http://travis-ci.org/CodementorIO/rest-firebase) [![Coverage Status](https://coveralls.io/repos/CodementorIO/rest-firebase/badge.png)](https://coveralls.io/r/CodementorIO/rest-firebase) [![Join the chat at https://gitter.im/CodementorIO/rest-firebase](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/CodementorIO/rest-firebase)

by [Codementor][]

[Codementor]: https://www.codementor.io/

## LINKS:

* [github](https://github.com/CodementorIO/rest-firebase)
* [rubygems](https://rubygems.org/gems/rest-firebase)
* [rdoc](http://rdoc.info/projects/CodementorIO/rest-firebase)

## DESCRIPTION:

Ruby Firebase REST API client built on top of [rest-core][].

[rest-core]: https://github.com/godfat/rest-core

## FEATURES:

* Concurrent requests
* Streaming requests

## REQUIREMENTS:

### Mandatory:

* Tested with MRI (official CRuby), Rubinius and JRuby.
* gem [rest-core][]
* gem [httpclient][]
* gem [mime-types][]
* gem [timers][]

[httpclient]: https://github.com/nahi/httpclient
[mime-types]: https://github.com/halostatue/mime-types
[timers]: https://github.com/celluloid/timers

### Optional:

* gem json or yajl-ruby, or multi_json

## INSTALLATION:

``` shell
gem install rest-firebase
```

Or if you want development version, put this in Gemfile:

``` ruby
gem 'rest-firebase', :git => 'git://github.com/CodementorIO/rest-firebase.git',
                     :submodules => true
```

## SYNOPSIS:

Check out Firebase's
[REST API documentation](https://www.firebase.com/docs/rest-api.html)
for a complete reference.

``` ruby
require 'rest-firebase'

f = RestFirebase.new :site => 'https://SampleChat.firebaseIO-demo.com/',
                     :secret => 'secret',
                     :d => {:auth_data => 'something'},
                     :log_method => method(:puts),
                     # `timeout` in seconds
                     :timeout => 10,
                     # `max_retries` upon failures. Default is: `0`
                     :max_retries => 3,
                     # `retry_exceptions` for which exceptions should retry
                     # Default is: `[IOError, SystemCallError]`
                     :retry_exceptions =>
                       [IOError, SystemCallError, Timeout::Error],
                     # `error_callback` would get called each time there's
                     # an exception. Useful for monitoring and logging.
                     :error_callback => method(:p),
                     # `auth_ttl` describes when we should refresh the auth
                     # token. Set it to `false` to disable auto-refreshing.
                     # The default is 23 hours.
                     :auth_ttl => 82800,
                     # `auth` is the auth token from Firebase. Leave it alone
                     # to auto-generate. Set it to `false` to disable it.
                     :auth => false # Ignore auth for this example!

@reconnect = true

# Streaming over 'users/tom'
es = f.event_source('users/tom')
es.onopen   { |sock| p sock } # Called when connected
es.onmessage{ |event, data, sock| p event, data } # Called for each message
es.onerror  { |error, sock| p error } # Called whenever there's an error
# Extra: If we return true in onreconnect callback, it would automatically
#        reconnect the node for us if disconnected.
es.onreconnect{ |error, sock| p error; @reconnect }

# Start making the request
es.start

# Try to close the connection and see it reconnects automatically
es.close

# Update users/tom.json
p f.put('users/tom', :some => 'data')
p f.post('users/tom', :some => 'other')
p f.get('users/tom')
p f.delete('users/tom')

# With Firebase queries (it would encode query in JSON for you)
p f.get('users/tom', :orderBy => '$key', :limitToFirst => 1)

# Need to tell onreconnect stops reconnecting, or even if we close
# the connection manually, it would still try to reconnect again.
@reconnect = false

# Close the connection to gracefully shut it down.
es.close

# Refresh the auth by resetting it
f.auth = nil
```

## Concurrent HTTP Requests:

Inherited from [rest-core][], you can do concurrent requests quite easily.
Here's a very quick example of making two API calls at the same time.

``` ruby
require 'rest-firebase'
firebase = RestFirebase.new(:log_method => method(:puts))
puts "httpclient with threads doing concurrent requests"
a = [firebase.get('users/tom'), firebase.get('users/mom')]
puts "It's not blocking... but doing concurrent requests underneath"
p a.map{ |r| r['name'] } # here we want the values, so it blocks here
puts "DONE"
```

If you prefer callback based solution, this would also work:

``` ruby
require 'rest-firebase'
firebase = RestFirebase.new(:log_method => method(:puts))
puts "callback also works"
firebase.get('users/tom') do |r|
  p r['name']
end
puts "It's not blocking... but doing concurrent requests underneath"
firebase.wait # we block here to wait for the request done
puts "DONE"
```

For a detailed explanation, see:
[Advanced Concurrent HTTP Requests -- Embrace the Future][future]

[future]: https://github.com/godfat/rest-core#advanced-concurrent-http-requests----embrace-the-future

### Thread Pool / Connection Pool

Underneath, rest-core would spawn a thread for each request, freeing you
from blocking. However, occasionally we would not want this behaviour,
giving that we might have limited resource and cannot maximize performance.

For example, maybe we could not afford so many threads running concurrently,
or the target server cannot accept so many concurrent connections. In those
cases, we would want to have limited concurrent threads or connections.

``` ruby
RestFirebase.pool_size = 10
RestFirebase.pool_idle_time = 60
```

This could set the thread pool size to 10, having a maximum of 10 threads
running together, growing from requests. Each threads idled more than 60
seconds would be shut down automatically.

Note that `pool_size` should at least be larger than 4, or it might be
very likely to have _deadlock_ if you're using nested callbacks and having
a large number of concurrent calls.

Also, setting `pool_size` to `-1` would mean we want to make blocking
requests, without spawning any threads. This might be useful for debugging.

### Gracefully shutdown

To shutdown gracefully, consider shutdown the thread pool (if we're using it),
and wait for all requests for a given client. For example:

``` ruby
RestFirebase.shutdown
```

We could put them in `at_exit` callback like this:

``` ruby
at_exit do
  RestFirebase.shutdown
end
```

If you're using unicorn, you probably want to put that in the config.

## Powered sites:

* [Codementor][]

## CHANGES:

* [CHANGES](CHANGES.md)

## CONTRIBUTORS:

* Lin Jen-Shin (@godfat)
* Yoshihiro Ibayashi (@chanibarin)

## LICENSE:

Apache License 2.0

Copyright (c) 2014-2015, Codementor

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   <http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGES.md
# CHANGES

## rest-firebase 1.1.0 -- 2016-02-04

* Adopted rest-core 4.0.0

## rest-firebase 1.0.3 -- 2015-11-16

### Bugs fixed

* Raise the default `max_redirects` from 1 to 5 because Firebase introduced
  an extra redirect for EventSource. 2 is enough in theory but to make it
  more future compatible, we set the default to 5 for now. If Firebase
  really needs more than 5 redirects, you could also workaround this by
  setting `max_redirects` to another number while setting up the client.
  For example:

  ``` ruby
  client = RestFirebase.new(:max_redirects => 10)
  # or
  client = RestFirebase.new
  client.max_redirects = 10
  ```

  This works for any version of rest-firebase. Thanks @chanibarin
  See: <https://github.com/CodementorIO/rest-firebase/pull/7>

## rest-firebase 1.0.2 -- 2015-06-12

* Fixed a bug where it would try to encode JSON twice upon retrying.

## rest-firebase 1.0.1 -- 2015-01-04

* Ruby 2.2 compatibility

## rest-firebase 1.0.0 -- 2014-12-09

### Enhancement

* Encode query in JSON to make using [Firebase queries][] easy.
* Introduced `max_retries`, `retry_exceptions`, and `error_callback` from
  latest rest-core (3.5.0+). See README.md for detail.

[Firebase queries]: https://www.firebase.com/docs/rest/guide/retrieving-data.html#section-rest-queries

### Internal Enhancement

* Encode payload in JSON with middleware from rest-core

## rest-firebase 0.9.5 -- 2014-11-07

* Base64url encoded JWT would no longer contain any newlines.

## rest-firebase 0.9.4 -- 2014-09-01

* Should really properly refresh the auth (query)
* From now on you're not allowed to change the value of query.

## rest-firebase 0.9.3 -- 2014-08-25

* Adopted rest-core 3.3.0
* Introduce `RestFirebase#auth_ttl` to setup when to refresh the auth token.
  Default to 23 hours (82800 seconds)
* Properly refresh the auth token by resetting `RestFirebase#iat`.

## rest-firebase 0.9.2 -- 2014-08-06

* Now it would auto-refresh auth if it's also expired (>= 23 hours)

## rest-firebase 0.9.1 -- 2014-06-28

* Now it would properly send JSON payload and headers.

## rest-firebase 0.9.0 -- 2014-05-13

* Birthday!


test/test_api.rb

require 'rest-firebase'
require 'rest-builder/test'

Pork::API.describe RestFirebase do
  before do
    stub_select_for_stringio
    stub(Time).now{ Time.at(86400) }
  end

  after do
    WebMock.reset!
    Muack.verify
  end

  path = 'https://a.json?auth=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ2IjowLCJpYXQiOjg2NDAwLCJkIjpudWxsfQ%3D%3D.SSmw2fUYiQFyYlsFV8WmyQsOCWJ6yvC7aw3bRpwQOYo%3D'

  json = '{"status":"ok"}'
  rbon = {'status' => 'ok'}

  def firebase
    @firebase ||= RestFirebase.new(:secret => 'nnf')
  end

  would 'get true' do
    stub_request(:get, path).to_return(:body => 'true')
    firebase.get('https://a').should.eq true
  end

  would 'get true with callback' do
    stub_request(:get, path).to_return(:body => 'true')
    firebase.get('https://a') do |r|
      r.should.eq true
    end.wait
  end

  would 'get with query' do
    stub_request(:get, "#{path}&orderBy=%22date%22&limitToFirst=1").
      to_return(:body => json)
    firebase.get('https://a', :orderBy => 'date', :limitToFirst => 1).
      should.eq rbon
  end

  would 'put {"status":"ok"}' do
    stub_request(:put, path).with(:body => json).to_return(:body => json)
    firebase.put('https://a', rbon).should.eq rbon
  end

  would 'have no payload for delete' do
    stub_request(:delete, path).with(:body => nil).to_return(:body => json)
    firebase.delete('https://a').should.eq rbon
  end

  would 'parse event source' do
    stub_request(:get, path).to_return(:body => <<-SSE)
event: put
data: {}

event: keep-alive
data: null

event: invalid
data: invalid
SSE
    m = [{'event' => 'put'       , 'data' => {}},
         {'event' => 'keep-alive', 'data' => nil}]
    es = firebase.event_source('https://a')
    es.should.kind_of? RestFirebase::Client::EventSource
    es.onmessage do |event, data|
      {'event' => event, 'data' => data}.should.eq m.shift
    end.onerror do |error|
      error.should.kind_of? RC::Json::ParseError
    end.start.wait
    m.should.empty?
  end

  would 'refresh token' do
    firebase # initialize http-client first (it's using Time.now too)
    mock(Time).now{ Time.at(0) }
    auth, query = firebase.auth, firebase.query
    query[:auth].should.eq auth
    Muack.verify(Time)
    stub(Time).now{ Time.at(86400) }

    stub_request(:get, path).to_return(:body => 'true')
    firebase.get('https://a').should.eq true
    firebase.auth .should.not.eq auth
    firebase.query.should.not.eq query
  end

  would 'not double encode json upon retrying' do
    stub_request(:post, path).
      to_return(:body => '{}', :status => 500).times(1).then.
      to_return(:body => '[]', :status => 200).times(1).
        with(:body => '{"is":"ok"}')

    firebase.retry_exceptions = [TrueClass]
    firebase.max_retries      = 1
    firebase.error_handler    = false
    expect(firebase.post(path, :is => :ok)).eq([])
  end

  define_method :check do |status, klass|
    stub_request(:delete, path).to_return(
      :body => '{}', :status => status)

    lambda{ firebase.delete('https://a').tap{} }.should.raise(klass)

    WebMock.reset!
  end

  would 'raise exception when encountering error' do
    [400, 401, 402, 403, 404, 406, 417].each do |status|
      check(status, RestFirebase::Error)
    end
    [500, 502, 503].each do |status|
      check(status, RestFirebase::Error::ServerError)
    end
  end
end


test/firebase/test_firebase.rb

require 'rest-more/test'

describe RC::Firebase do
  before do
    stub(Time).now{ Time.at(0) }
  end

  after do
    WebMock.reset!
    Muack.verify
  end

  path = 'https://a.json?auth=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9%0A.eyJ2IjowLCJpYXQiOjAsImQiOm51bGx9%0A.C9JtzZhiCrsClNdAQcE7Irngr2BZJCH4x1p-IHxfrAo%3D%0A'

  def firebase
    RC::Firebase.new(:secret => 'nnf')
  end

  should 'get true' do
    stub_request(:get, path).to_return(:body => 'true')
    firebase.get('https://a').should.eq true
  end

  should 'put {"status":"ok"}' do
    json = '{"status":"ok"}'
    rbon = {'status' => 'ok'}
    stub_request(:put, path).with(:body => json).to_return(:body => json)
    firebase.put('https://a', rbon).should.eq rbon
  end

  should 'parse event source' do
    stub_request(:get, path).to_return(:body => <<-SSE)
event: put
data: {}

event: keep-alive
data: null

event: invalid
data: invalid
SSE
    m = [{'event' => 'put'       , 'data' => {}},
         {'event' => 'keep-alive', 'data' => nil}]
    es = firebase.event_source('https://a')
    es.should.kind_of RC::Firebase::Client::EventSource
    es.onmessage do |event, data|
      {'event' => event, 'data' => data}.should.eq m.shift
    end.onerror do |error|
      error.should.kind_of RC::Json::ParseError
    end.start.wait
    m.should.empty
  end

  check = lambda do |status, klass|
    stub_request(:delete, path).to_return(
      :body => '{}', :status => status)

    lambda{ firebase.delete('https://a').tap{} }.should.raise(klass)

    WebMock.reset!
  end

  should 'raise exception when encountering error' do
    [400, 401, 402, 403, 404, 406, 417].each do |status|
      check[status, RC::Firebase::Error]
    end
    [500, 502, 503].each do |status|
      check[status, RC::Firebase::Error::ServerError]
    end
  end
end


lib/rest-core/client/firebase.rb

require 'rest-core'

# https://www.firebase.com/docs/security/custom-login.html
# https://www.firebase.com/docs/rest-api.html
module RestCore
  Firebase = Builder.client(:d, :secret, :auth) do
    use Timeout       , 10

    use DefaultSite   , 'https://SampleChat.firebaseIO-demo.com/'
    use DefaultHeaders, {'Accept' => 'application/json'}
    use DefaultQuery  , nil

    use FollowRedirect, 1
    use CommonLogger  , nil
    use Cache         , nil, 600 do
      use ErrorHandler, lambda{ |env| Firebase::Error.call(env) }
      use ErrorDetectorHttp
      use JsonResponse, true
    end
  end
end

class RestCore::Firebase::Error < RestCore::Error
  include RestCore
  class ServerError         < Firebase::Error; end
  class ClientError         < RestCore::Error; end

  class BadRequest          < Firebase::Error; end
  class Unauthorized        < Firebase::Error; end
  class Forbidden           < Firebase::Error; end
  class NotFound            < Firebase::Error; end
  class NotAcceptable       < Firebase::Error; end
  class ExpectationFailed   < Firebase::Error; end

  class InternalServerError < Firebase::Error::ServerError; end
  class BadGateway          < Firebase::Error::ServerError; end
  class ServiceUnavailable  < Firebase::Error::ServerError; end

  attr_reader :error, :code, :url
  def initialize error, code, url=''
    @error, @code, @url = error, code, url
    super("[#{code}] #{error.inspect} from #{url}")
  end

  def self.call env
    error, code, url = env[RESPONSE_BODY], env[RESPONSE_STATUS],
                       env[REQUEST_URI]
    return new(error, code, url) unless error.kind_of?(Hash)
    case code
      when 400; BadRequest
      when 401; Unauthorized
      when 403; Forbidden
      when 404; NotFound
      when 406; NotAcceptable
      when 417; ExpectationFailed
      when 500; InternalServerError
      when 502; BadGateway
      when 503; ServiceUnavailable
      else    ; self
    end.new(error, code, url)
  end
end

module RestCore::Firebase::Client
  include RestCore

  class EventSource < RestCore::EventSource
    def onmessage event=nil, data=nil, sock=nil
      if event
        super(event, Json.decode(data), sock)
      else
        super
      end
    end
  end

  def request env, app=app
    super(env.merge(REQUEST_PATH    => "#{env[REQUEST_PATH]}.json",
                    REQUEST_PAYLOAD => Json.encode(env[REQUEST_PAYLOAD])),
          app)
  end

  def generate_auth opts={}
    raise Firebase::Error::ClientError.new(
      "Please set your secret") unless secret

    header = {:typ => 'JWT', :alg => 'HS256'}
    claims = {:v => 0, :iat => Time.now.to_i, :d => d}.merge(opts)
    # http://tools.ietf.org/html/draft-ietf-jose-json-web-signature-26
    input = [header, claims].map{ |d| base64url(Json.encode(d)) }.join('.')
    # http://tools.ietf.org/html/draft-ietf-oauth-json-web-token-20
    "#{input}.#{base64url(Hmac.sha256(secret, input))}"
  end

  private
  def base64url str; [str].pack('m').tr('+/', '-_'); end
  def default_query; {:auth => auth}; end
  def default_auth ; generate_auth  ; end
end

class RestCore::Firebase
  include RestCore::Firebase::Client
  self.event_source_class = EventSource
end


lib/rest-firebase.rb

require 'rest-core'
require 'rest-core/util/json'
require 'rest-core/util/hmac'

# https://www.firebase.com/docs/security/custom-login.html
# https://www.firebase.com/docs/rest-api.html
# https://www.firebase.com/docs/rest/guide/retrieving-data.html#section-rest-queries
RestFirebase =
  RestCore::Builder.client(:d, :secret, :auth, :auth_ttl, :iat) do
    use RestCore::DefaultSite   , 'https://SampleChat.firebaseIO-demo.com/'
    use RestCore::DefaultHeaders, {'Accept' => 'application/json',
                             'Content-Type' => 'application/json'}
    use RestCore::DefaultQuery  , nil
    use RestCore::JsonRequest   , true

    use RestCore::Retry         , 0, RestCore::Retry::DefaultRetryExceptions
    use RestCore::Timeout       , 10
    use RestCore::FollowRedirect, 5
    use RestCore::ErrorHandler  , lambda{|env| RestFirebase::Error.call(env)}
    use RestCore::ErrorDetectorHttp
    use RestCore::JsonResponse  , true
    use RestCore::CommonLogger  , nil
    use RestCore::Cache         , nil, 600
  end

class RestFirebase::Error < RestCore::Error
  class ServerError         < RestFirebase::Error; end
  class ClientError         < RestCore::Error; end

  class BadRequest          < RestFirebase::Error; end
  class Unauthorized        < RestFirebase::Error; end
  class Forbidden           < RestFirebase::Error; end
  class NotFound            < RestFirebase::Error; end
  class NotAcceptable       < RestFirebase::Error; end
  class ExpectationFailed   < RestFirebase::Error; end

  class InternalServerError < RestFirebase::Error::ServerError; end
  class BadGateway          < RestFirebase::Error::ServerError; end
  class ServiceUnavailable  < RestFirebase::Error::ServerError; end

  attr_reader :error, :code, :url
  def initialize error, code, url=''
    @error, @code, @url = error, code, url
    super("[#{code}] #{error.inspect} from #{url}")
  end

  def self.call env
    error, code, url = env[RestCore::RESPONSE_BODY],
                       env[RestCore::RESPONSE_STATUS],
                       env[RestCore::REQUEST_URI]
    return new(error, code, url) unless error.kind_of?(Hash)
    case code
      when 400; BadRequest
      when 401; Unauthorized
      when 403; Forbidden
      when 404; NotFound
      when 406; NotAcceptable
      when 417; ExpectationFailed
      when 500; InternalServerError
      when 502; BadGateway
      when 503; ServiceUnavailable
      else    ; self
    end.new(error, code, url)
  end
end

module RestFirebase::Client
  class EventSource < RestCore::EventSource
    def onmessage event=nil, data=nil, sock=nil
      if event
        super(event, RestCore::Json.decode(data), sock)
      else
        super
      end
    end
  end

  def request env, a=app
    check_auth
    query = env[RestCore::REQUEST_QUERY].inject({}) do |q, (k, v)|
      q[k] = RestCore::Json.encode(v)
      q
    end
    super(env.merge(RestCore::REQUEST_PATH =>
                      "#{env[RestCore::REQUEST_PATH]}.json",
                    RestCore::REQUEST_QUERY => query), a)
  end

  def generate_auth opts={}
    raise RestFirebase::Error::ClientError.new(
      "Please set your secret") unless secret

    self.iat = nil
    header = {:typ => 'JWT', :alg => 'HS256'}
    claims = {:v => 0, :iat => iat, :d => d}.merge(opts)
    # http://tools.ietf.org/html/draft-ietf-jose-json-web-signature-26
    input = [header, claims].map{ |d| base64url(RestCore::Json.encode(d)) }.
            join('.')
    # http://tools.ietf.org/html/draft-ietf-oauth-json-web-token-20
    "#{input}.#{base64url(RestCore::Hmac.sha256(secret, input))}"
  end

  def query
    {:auth => auth}
  end

  private
  def base64url str; [str].pack('m0').tr('+/', '-_'); end
  def default_auth    ; generate_auth  ; end
  def default_auth_ttl; 82800          ; end
  def default_iat     ; Time.now.to_i  ; end

  def check_auth
    self.auth = nil if auth_ttl && Time.now.to_i - iat > auth_ttl
  end
end

class RestFirebase
  include RestFirebase::Client
  self.event_source_class = EventSource
  const_get(:Struct).send(:remove_method, :query=)
end
